File: python/problem2.py
#!/usr/bin/env python2.5

"""
Problem 2 from Project Euler

Find the sum of all the even-valued terms in the sequence which do not exceed four million.
"""

def fib(n):
    """
    Fibonacci sequence n
    """
    if n >= 3:
        f = [1, 2]
        for i in range(1, n-1):
            fa = f[i] + f[i-1]
            f.append(fa)
            
        return f
        
def fibnum(n):
    """
    Fibonnaci number for sequence n
    """
    return fib(n)[-1]

fmax = 4000000
fibt = 0
fibmax = 0

for i in range(10, 50):
    if fibnum(i) <= fmax:
        fibmax = i
    else:
        break
    
for j in fib(fibmax):
    if j % 2 == 0:
        fibt += j
        
print fibt

File: go/problem2.go
package main

/*
 *  Problem 2 from Project Euler
 *
 *  By considering the terms in the Fibonacci sequence whose values do not 
 *  exceed four million, find the sum of the even-valued terms.
 */

const N = 4000000

func main() {
	fib := fibonacci()
	i := fib()
	sum := 0

	for i < N {
		if i%2 == 0 {
			sum += i
		}
		i = fib()
	}

	println(sum)
}

func fibonacci() func() int {
	var f1, f2 = 0, 1
	f := func() int {
		f0 := f1
		f1, f2 = f2, f1+f2
		return f0
	}
	return f
}
